templates/cmake.mpd

```diff
@@ -118,6 +118,10 @@ add_library(${PROJECT_TARGET_<%uc(normalize(project_name))%>} ${SOURCE_FILES_<%u
 <%if(staticflags)%>
 target_compile_definitions(${PROJECT_TARGET_<%uc(normalize(project_name))%>} PRIVATE <%staticflags%>)
 <%endif%>
+<%else%>
+<%if(custom_types)%>
+add_library(${PROJECT_TARGET_<%uc(normalize(project_name))%>} OBJECT)
+<%endif%>
```

Comment:
What does this do and when does it come into play?

Reply:
Two things.

First, declaring a compilable target is (currently) required because MPC (currently) emits CMake code that sets several properties of the target `${PROJECT_TARGET_<%uc(normalize(project_name))%>}`, including via `set_target_properties()`, `target_compile_definitions()`, and `target_link_directories()`.  And currently, no target is declared for custom_types with no library or executable (e.g. IDL).  So declaring an object library for custom targets was the smallest change that allowed me to build ACE+TAO.  However, I'm not sure if that's "correct" as I don't know the full scope of what `custom_types` are.

* Instead of declaring an object library target, you could conceivably add logic to disable setting those properties for `custom_types`, but like I said, I don't know what `custom_types` could be.  Perhaps there are `custom_types` whose output you'd want to compile & link into a library?  (I collect my IDL outputs into static or object libraries.)

Second: if `tao_idl` outputs CPP/H files into the source directory (where all TAO projects expect them), then it's necessary to add those files to _some target_ in the same directory as the CMakeLists.txt that calls `tao_idl`.  (In this case, for simplicity `IDL_FILES_TARGET_SOURCES()` adds them to the main target of the very file that calls it).  If you don't, then CMake does not detect that a command (`tao_idl`) exists to resolve dependencies on them, and CMake will report missing dependency errors for projects that use them.  This prominently happens for `AnyTypeCode` and `PortableServer`

* Conceivably, `IDL_FILES_TARGET_SOURCES()` could declare new object library targets for each IDL file, perhaps based on the IDL filename without an extension.

Reply:
I'm open to recommendations or edits for both points.

Reply:
Does it require the project that is only a custom_type to create a library?  If not, then I'm ok with this change.

Reply:
I'm not sure entirely what you mean.

Currently, that `custom_types` object library target is only emitted if the project did not have an executable or (shared or static) library.  Do you know whether `custom_types` projects can ever have those?

Reply:
I'm asking, what does this line do for the project?  In MPC, custom_type only projects do not create libraries.

Reply:
@alexchandel it may help to define these terms.  See MPC's README and html docs for more details.

`custom_types` is a data structure that can be used in MPC templates.  It contains the details of the `Define_Custom` sections of the input.  Typically these are "custom build steps" that run commands that are not compiling/linking to generate source files or inputs to other custom build steps.

`custom_only` is a kind of MPC project that MSVS calls "Utility" -- it doesn't produce object files, a library or an executable, only runs custom build steps.

Reply:
@mitza-oci This helps.  Important question: is it possible to have a `custom_types` project that is not `custom_only`?  No MPC files in ACE/TAO result in CMakeLists project that has a `custom_types` plus any other targets (`exename`, `sharedname`, or `staticname`) that I found, but is it _possible_?

Reply:
@ocielliottc I've changed the MPC implementation so that custom_types no longer require an object library (and rebased/squashed commits).

The new [IDL macro implementation](https://github.com/alexchandel/test-cmake-acetao/blob/master/cmake/InitAceTao.cmake) creates a subtarget for individual IDL files, and adds them to a project target (an exe or library) if one exists.